src/Data/Witherable.purs

```diff
@@ -97,16 +103,78 @@ withered :: forall t m x. Witherable t => Applicative m =>
 withered = wither identity
 
 instance witherableArray :: Witherable Array where
-  wilt = wiltDefault
-  wither = witherDefault
+  wilt p xs = ado
+    xs' <- traverse p xs
+    let left = ST.run do
+         ls <- STA.empty
+         iter <- STAI.iterator (xs' !! _)
+         STAI.iterate iter case _ of
+           Left l -> void $ STA.push l ls
+           _      -> pure unit
 
-instance witherableList :: Witherable List where
-  wilt = wiltDefault
-  wither = witherDefault
+         STA.unsafeFreeze ls
+
+    let right = ST.run do
+         rs <- STA.empty
+         iter  <- STAI.iterator (xs' !! _)
+         STAI.iterate iter case _ of
+           Right r -> void $ STA.push r rs
+           _       -> pure unit
+
+         STA.unsafeFreeze rs
+
+    in { left, right }
```

Comment:
Two comments on this.

Firstly:
Couldn't we build up both `ls` and `rs` in the same `ST.run` and then at the end:
```
{left: _, right: _} <$> STA.unsafeFreeze ls <*> STA.unsafeFreeze rs
```

Secondly: This is a reimplementation of `separate` from `Compactable`.  I'd prefer to move this implementation there, and then use `wiltDefault` here (`map separate <<< traverse p`).

Reply:
> This is a reimplementation of `separate` from `Compactable`

I was too deep in the implementation to realize this 😊 

I can certainly move it. Of course this means that I've rewritten `compact` in my `wither` implementation. Looking at how `catMaybes` is implemented leads me to wonder if this version might be faster.

> Couldn't we build up both `ls` and `rs` in the same `ST.run`

This is what I had originally attempted. Unless I'm mistaken, these two `STArray`s have different `Region`s and therefore cannot be run together. I _think_ that's the source of the error I spoke of above. I can certainly try again though.